### ISSUE TITLE
Project SDK not used when compiling Java code

### ISSUE THREAD

**Issue by hplatou** [cursive-ide/cursive#2428]
When including ':java-source-paths' in project.clj and using a Project SDK (in my case Java  8) I get the following problem when starting repl:

```
#error {
 :cause org/some/util/SomeJavaClass has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

It seems that the Java source is compiled using the Java version available on system path rather than using the one specified wihtin Project SDK.

Cursive: 1.9.4-eap3-2020.2
IntelliJ: 2020.2.1

**Comment by cursive-ide** [comment]
Interesting, I'm not entirely sure why that is happening, it's possible that the Java source is being compiled using the JDK used by IntelliJ itself - I'm not sure at which stage lein compiles Java classes.

Does adding this option to your `project.clj` help?
```
:javac-options ["-target" "1.8" "-source" "1.8"]
```

**Comment by hplatou** [comment]
Yes! Adding this option makes it compile. I guess it is still not using the project SDK, but at least the class files are compatible. This resolves my problem. Thanks!

**Comment by cursive-ide** [comment]
Exactly, I'd have to investigate to fully figure it out, but I suspect that your IntelliJ JDK (probably 11) is doing the compiling somehow. But I'm glad the workaround works around!

**Comment by d-t-w** [comment]
Hey @cursive-ide I think this is a Leiningen default JDK resolution issue where `JAVA_HOME` and `JAVA_CMD` are not set.

TLDR; I think you can close this ticket down, users should either:

1. Set `JAVA_HOME` or `JAVA_CMD` (see related [Leiningen ticket](https://github.com/technomancy/leiningen/issues/2228)), or
2. Set explicit `:javac-options ["--release" "<desired javac version>"]` or similar

---

In my local setup I have a lot of JDKs (from 8 up to 21).

I use [Jenv](https://www.jenv.be/) to manage those versions locally. I currently have it set to JDK 17.

```
❯ jenv version
17 (set by /Users/derek/.jenv/version)

❯ java --version
openjdk 17.0.13 2024-10-15
OpenJDK Runtime Environment Temurin-17.0.13+11 (build 17.0.13+11)
OpenJDK 64-Bit Server VM Temurin-17.0.13+11 (build 17.0.13+11, mixed mode, sharing)
```


I have my Intellij project SDK set to Java 17.

<img width="1085" alt="Screenshot 2024-12-06 at 12 34 05 PM" src="https://github.com/user-attachments/assets/5b6955a1-0cb1-4ffb-9c25-81b49aaf7ee0">


In my Intellij Terminal I can see the correct SDK is used:

<img width="739" alt="Screenshot 2024-12-06 at 12 34 33 PM" src="https://github.com/user-attachments/assets/f657dffd-cc56-43cd-9293-a2560ed6a9b3">


And yet.. when I open the REPL I get a bytecode version mismatch between compilation (JDK21) and runtime (JDK17)

```
/Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home/bin/java <..snip..> Compiling kpow.kafka.serdes
Syntax error macroexpanding at (avro.clj:1:1).
Execution error (UnsupportedClassVersionError) at java.lang.ClassLoader/defineClass1 (ClassLoader.java:-2).
io/kpow/avro/KpowDecimalConversion has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0
```


Jenv does not set `JAVA_HOME` by default, [you have to activate a plugin](https://github.com/jenv/jenv/issues/44).

With no `JDK_HOME` or `JDK_CMD` set, Lein appears to be somehow using the JDK installed at `/usr/bin/java` which in my case just happens to be JDK21.

```
/usr/bin/java --version
openjdk 21.0.5 2024-10-15 LTS
OpenJDK Runtime Environment Temurin-21.0.5+11 (build 21.0.5+11-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.5+11 (build 21.0.5+11-LTS, mixed mode, sharing)
```

**Comment by cursive-ide** [comment]
Thanks @d-t-w, I agree, I'll close this. Thanks for all the detail!
